### PR TITLE
assign Network Contributor role to AKS identities

### LINKED ISSUE #102: [AVM Module Issue]: UAMI Role Assignment for Network Related Actions Missing
### Check for previous/existing GitHub issues

- [X] I have checked for previous/existing GitHub issues

### Issue Type?

Bug

### (Optional) Module Version

_No response_

### (Optional) Correlation Id

_No response_

### Description

If using bring your own network, UAMI, and deploying using IaC (non-AZ CLI), the AKS docs state that: `If you're using an ARM template or another method, use the principal ID of the managed identity to perform a role assignment. ` 

Goes on further to say: `For a VNet, attached Azure disk, static IP address, or route table outside the default worker node resource group, you need to assign the Network Contributor role on the custom resource group.` 

Ref: [Add a role assignment for a system-assigned managed identity](https://learn.microsoft.com/en-us/azure/aks/use-managed-identity#assign-an-azure-rbac-role-to-the-system-assigned-managed-identity)  

If a user forgets to do this, the future deployment of K8s network related resources like Istio External Gateway or NGINX ingress with the application routing add-on, will end up in a failed state due to missing RBAC permissions. 

I recommend an update to this module to either:
- Mention this in the documentation of the module as a note/warning/reminder
- Add the role assignment for Network Contributor

### PR DESCRIPTION
**Pull request by zioproto**
## Description

Add necessary role assignment to fix #102

Closes #102 

## Type of Change

- [ ] Non-module change (e.g. CI/CD, documentation, etc.)
- [ ] Azure Verified Module updates:
  - [ X] Bugfix containing backwards compatible bug fixes
    - [X] Someone has opened a bug report issue, and I have included "Closes #{bug_report_issue_number}" in the PR description.
    - [ ] The bug was found by the module author, and no one has opened an issue to report it yet.
  - [ ] Feature update backwards compatible feature updates.
  - [ ] Breaking changes.
  - [ ] Update to documentation

### CODE REVIEW

**Comment by senavar** [comment]
Hi @zioproto, I think the network contributor role needs to be scoped for the entire RG to allow for greater flexibility. For example, if using user-assigned NAT gateway the UAMI needs permission to do action upon the NAT Gateway when deploying Application add-on. Here is an example where the UAMI has access to the vNet but not the NAT GW: 
image

File: main.tf [hunk 1]
```diff
@@ -29,6 +29,23 @@ resource "azurerm_user_assigned_identity" "aks" {
...
+  # /subscriptions/00000000-0000-0000-0000-000000000000/resourcegroups/acceptanceTestResourceGroup1/providers/Microsoft.ManagedIdentity/userAssignedIdentities/testIdentity
+  # name is obtained from the above string which is the user assigned resource id - reference https://github.com/Azure/terraform-azurerm-aks/blob/decb533e2785f965673698b0ac9949faca963f68/role_assignments.tf#L11
+  name                = split("/", tolist(var.managed_identities.user_assigned_resource_ids)[count.index])[8]
+  resource_group_name = split("/", tolist(var.managed_identities.user_assigned_resource_ids)[count.index])[4]
+}
+
+resource "azurerm_role_assignment" "network_contributor_on_subnet" {
+  // Use the principal_id from the user assigned identity if it exists, otherwise use the principal_id from the AKS cluster
+  // reference https://github.com/Azure/terraform-azurerm-aks/blob/decb533e2785f965673698b0ac9949faca963f68/role_assignments.tf#L27
+  principal_id         = coalesce(try(data.azurerm_user_assigned_identity.cluster_user_defined_identity[0].principal_id, azurerm_kubernetes_cluster.this.identity[0].principal_id))
```
Comment (nellyk) [inline main.tf:44]: without and with availability zones examples use ```data.azurerm_user_assigned_identity.cluster_user_defined_identity[0].principal_id```
whilst azurerm_kubernetes_cluster.this.identity[0].principal_id) uses the generated identity

File: main.tf [hunk 2]
```diff
@@ -29,6 +29,23 @@ resource "azurerm_user_assigned_identity" "aks" {
   tags                = var.tags
 }
 
+data "azurerm_user_assigned_identity" "cluster_user_defined_identity" {
+  count = length(var.managed_identities.user_assigned_resource_ids) > 0 ? length(var.managed_identities.user_assigned_resource_ids) : 0
```
Comment (zioproto) [inline main.tf:33]:
```suggestion
  count = length(var.managed_identities.user_assigned_resource_ids) > 0 ? 1 : 0
```

Comment (zioproto) [reply main.tf:33]: Did we talk about assumption that the identity is only 1 ?

Comment (nellyk) [reply main.tf:33]: we did and decided for it to be only one, let me remove this

File: main.tf [hunk 3]
```diff
@@ -29,6 +29,20 @@ resource "azurerm_user_assigned_identity" "aks" {
...
+
+
+data "azurerm_resource_group" "this" {
+  name = var.resource_group_name
+}
+
+resource "azurerm_role_assignment" "network_contributor_on_resource_group" {
+  #  Use the principal_id from the user assigned identity if it exists, otherwise use the principal_id from the AKS cluster
+  #  reference https://github.com/Azure/terraform-azurerm-aks/blob/decb533e2785f965673698b0ac9949faca963f68/role_assignments.tf#L27
+  principal_id         = azurerm_kubernetes_cluster.this.identity[0].principal_id
```
[the remark this answers is not in the document]
Comment (nellyk) [reply main.tf]: Thanks for catching this, it seems for user assigned managed identity the block is `identities` as per your commit

**Review by nellyk — APPROVED** [review]

**Pull request merged**